### PR TITLE
[8.1/n] support TLS modes

### PR DESCRIPTION
**Pull request by jronak**
Define supported modes of TLS listener:
* permissive: accept TLS and plaintext connections
* enforced: accept TLS connections only. 

Integrate the modes into the tlsmux listener to reject plaintext connections in enforced mode.

### CODE REVIEW

File: api/transport/inboundtlsmode.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+var (
+	_ fmt.Stringer             = (*InboundTLSMode)(nil)
+	_ encoding.TextUnmarshaler = (*InboundTLSMode)(nil)
+)
+
+// InboundTLSMode represents the TLS mode of the inbound.
+type InboundTLSMode uint16
+
+// String returns a lower-case ASCII representation of TLS mode.
+func (t InboundTLSMode) String() string {
```
Comment (AllenLuUber) [inline api/transport/inboundtlsmode.go:48]: sounds this should be part of stringer generation: https://pkg.go.dev/golang.org/x/tools/cmd/stringer

File: transport/internal/tlsmux/listener.go [hunk 2]
```diff
@@ -184,9 +192,14 @@ func (l *listener) handleTLSConn(conn net.Conn) (net.Conn, error) {
 	return tlsConn, nil
 }
 
-func (l *listener) handlePlaintextConn(conn net.Conn) net.Conn {
+func (l *listener) handlePlaintextConn(conn net.Conn) (net.Conn, error) {
+	if l.mode == transport.Enforced {
+		l.logger.Error("plaintext connection not allowed in enforced TLS mode: rejecting connection")
+		l.observer.incPlaintextConnectionRejects()
+		return nil, errors.New("plaintext connection not allowed in enforced TLS mode")
```
Comment (AllenLuUber) [inline transport/internal/tlsmux/listener.go:199]: nit: move the error to a top level var

**Review by AllenLuUber — APPROVED** [review]

File: transport/internal/tlsmux/listener_test.go [hunk 3]
```diff
@@ -120,28 +145,13 @@ func TestMux(t *testing.T) {
...
-				},
-				MinVersion: tls.VersionTLS10,
-				MaxVersion: tls.VersionTLS11,
-				RootCAs:    scenario.CAs,
-			},
-			expectError:    true,
-			clientErrorMsg: "remote error: tls: protocol version not supported",
+			clientErrorMsgOnDial:   "remote error: tls: protocol version not supported",
+			serverErrorMsgOnAccept: "listener closed",
+			mode:                   transport.Permissive,
```
Comment (gopalchouhan) [inline transport/internal/tlsmux/listener_test.go:154]: Can you please also add a test case for disabled mode?

File: api/transport/inboundtlsmode.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+const (
+	// Disabled TLS mode accepts plaintext connections only.
+	Disabled InboundTLSMode = iota
+
+	// Permissive TLS mode inbound accepts both TLS and plaintext connections.
+	Permissive
+
+	// Enforced TLS mode inbound accepts TLS connections only.
+	Enforced
```
[the remark this answers is not in the document]
Comment (jronak) [reply api/transport/inboundtlsmode.go]: This cannot be part of tlsmux package as it is an internal package. I have moved it to `api/transport/tls` and the imports now follow `yarpctls`

File: transport/internal/tlsmux/listener.go [hunk 5]
```diff
@@ -164,7 +172,7 @@ func (l *listener) mux(conn net.Conn) (net.Conn, error) {
 		return l.handleTLSConn(c)
```
[the remark this answers is not in the document]
Comment (gopalchouhan) [reply transport/internal/tlsmux/listener.go:176]: Makes sense! My only concern is: the way we handle the non tls traffic in enforced mode by rejecting it gracefully, wondering we do the same with existing listener for tls traffic in disabled mode. If you confirm that, I am fine with this setup.

**Review by gopalchouhan — APPROVED** [review]

**Pull request merged**